lib/pbench/server/sync.py

```diff
@@ -27,116 +19,156 @@ def __str__(self) -> str:
-        Advertise the operations for which the dataset is now ready.
-
-        TODO: It'd be nice to nest this inside a transaction so it's all
-        atomic. This seems to be difficult to achieve in a manner that works
-        both with PostgreSQL for production and sqlite3 for unit tests. When
-        we've removed the use of sqlite3 we can make this a nested transaction.
+        """Advertise the operations for which the dataset is now ready.
 
         Args:
             dataset: The dataset
-            did: The operation (if any) just completed, which will be removed
-                from the list of eligible operations.
+            did: The new OperationState of the component Operation
             enabled: A list (if any) of operations for which the dataset is now
                 eligible.
-            status: A status message (if not specified, and enabling new
-                operation(s), the default is "ok")
+            message: An optional status message
         """
-        ops: JSONVALUE = Metadata.getvalue(dataset, Metadata.OPERATION)
-        message = status
-
-        if not ops:
-            operations = set()
-        else:
-            operations = set(ops)
-            if did:
-                operations.discard(did.name)
-
-        # If we "did" something, the default message is "ok"
-        if did and not message:
-            message = "ok"
 
+        if enabled is None:
+            enabled = []
+
+        self.logger.debug(
+            "Dataset {} did {}, enabling {} with message {!r}",
+            dataset.name,
+            did.name if did else "nothing",
+            [e.name for e in enabled] if enabled else "none",
+            message,
+        )
+
+        # Don't reference the Dataset (which is outside our session scope)
+        # inside the context manager.
+        ds_name = dataset.name
+        ds_id = dataset.id
+        retries = self.RETRIES
+
+        # To avoid SELECT statements after updates, and the whole complication
+        # of SQLAlchemy AUTOFLUSH semantics, we're going to gather a set of all
+        # the operation rows we might need to update at the beginning.
+        match_set: set[OperationName] = set()
+        if did or message:
+            match_set.add(self.component)
         if enabled:
-            operations.update(o.name for o in enabled)
-
-        try:
-            self.logger.info(
-                "Did {}, enabling {} with message {!r}",
-                did.name if did else "nothing",
-                [e.name for e in enabled] if enabled else "none",
-                message,
-            )
-            Metadata.setvalue(dataset, Metadata.OPERATION, sorted(operations))
-            if message:
-                Metadata.setvalue(dataset, "server.status." + self.component, message)
-        except Exception as e:
-            self.logger.warning("{} error updating ops: {}", dataset.name, str(e))
-            raise
+            match_set.update(enabled)
```

Comment:
I'd actually rather have the `if`, because the intent it more clear, and it's more consistent with the other. Code's not necessarily better just because we drop a few characters.